enzyme/Enzyme/BlasDerivatives.td

```diff
@@ -117,81 +147,103 @@ def copy : CallBlasPattern<(Op $n, $x, $incx, $y, $incy),
 //                   ]
 //                   >;
 
-def asum : CallBlasPattern<(Op $n, $x, $incx),
-                  [],[len, vinc],
-                  [(noop)]// sum(abs(x))
-                  >;
-
+// def asum : CallBlasPattern<(Op $n, $x, $incx),
+//                   [],[len, vinc],
+//                   [(b<"axpy"> DiffeRet<"">, sign(<input<"x">), $y)]
+//                   >;
+// 
 
 // Lv 2
 
 // y = alpha Ax + beta y
-// def gemv : CallBlasPattern<(Op $layout, $transa, $m, $n, $alpha, $A, $lda, $x, $incx, $beta, $y, $incy),
-//                   [cblas_layout, trans, len, len, fp, mld, vinc, fp, vinc],
-//                   [
-//                   /* alpha  */  (b<"dot"> DiffeRet<"">, (b<"gemv"> $A, $x)),
-//                   /* A      */  //(),
-//                   /* x      */  //(),
-//                   /* beta   */  (b<"dot"> $y, DiffeRet<"">),
-//                   /* y      */  (b<"scal"> $beta, DiffeRet<"">),
-//                   ]
-//                   >;
+// currently assumes for vector dimensions that transa = 'N' and gets dimensions wrong otherwise
+def gemv : CallBlasPattern<(Op $layout, $transa, $m, $n, $alpha, $A, $lda, $x, $incx, $beta, $y, $incy),
+                  ["y"], [cblas_layout, trans,  len, len,  fp,   mld,      vinc,      fp,    vinc],
+                  [
+                  /* alpha  */  (inactive), //(b<"dot"> $y, (b<"gemv"> $A, $x)),
+                  /* A      */  (b<"ger">  $layout, $m, $n, $alpha, adj<"y">, $incy, $x, $incx, adj<"A">, $lda),
+                  /* x      */  (b<"gemv"> $layout, transpose<"transa">, $m, $n, $alpha, $A, (ld $A, $transa, $lda, $m, $n), adj<"y">, $incy, Constant<"1.0">, adj<"x">, $incx),
+                  /* beta   */  (inactive), //(b<"dot"> $y, DiffeRet<"">),
+                  /* y      */  (b<"scal"> (Rows $transa, $m, $n), $beta, adj<"y">, $incy)
+                  ]
+                  >;
 // 
-// def ger : CallBlasPattern<(Op $layout, $m, $n, $alpha, $x, $incx, $y, $incy, $v, $ld),
-//                   [cblas_layout, len, len, fp, vinc, vinc, vld],
-//                   []
-//                   >;
+def ger : CallBlasPattern<(Op $layout, $m, $n, $alpha, $x, $incx, $y, $incy, $A, $lda),
+                  ["A"],[cblas_layout, len, len, fp, vinc, vinc, mld],
+                  [
+                  (inactive),
+                  (inactive),
+                  (inactive),
+                  (inactive)
+                  ]
+                  >;
+//(ld $A, $transa, $lda, $m, $k)
```

Comment:
Should the entire ger pattern be copied out?

Reply:
No, because I do make use of the declaration. Otherwise we e.g. couldn't call it.